### ISSUE TITLE
3.9-Pytorch Loss Function and Optimizer

### ISSUE THREAD

**Issue by Pin-Jiun** [Pin-Jiun/Machine-Learing-HYL#24]
先看loss function怎麼建立並且使用backward計算微分

```py
import torch
import torchvision
from torch import nn
from torch.utils.data import DataLoader
# from torch.utils.tensorboard import SummaryWriter

dataset = torchvision.datasets.CIFAR10("/CIFAR10", train=False, transform=torchvision.transforms.ToTensor(), download=True)

dataloader = DataLoader(dataset, batch_size=64)

class PJL_NN(nn.Module):
    def __init__(self):
        super(PJL_NN,self).__init__()
        self.model = nn.Sequential(
            nn.Conv2d(3,32,5,padding=2),
            nn.MaxPool2d(2),
            nn.Conv2d(32,32,5,padding=2),
            nn.MaxPool2d(2),
            nn.Conv2d(32,64,5,padding=2),
            nn.MaxPool2d(2),
            nn.Flatten(),
            nn.Linear(1024,64),
            nn.Linear(64,10)
        )

    def forward(self, x):
        x = self.model(x)
        return x

#定義你所用的loss function, 關防有提供許多種loss 給你選
loss_fn = nn.CrossEntropyLoss()

jiun_nn = PJL_NN()

for data in dataloader:
    imgs, targets = data
    outputs = jiun_nn(imgs)
    loss = loss_fn(outputs, targets)

    #此時會更新jiun_nn裡面的nn每層的grad
    loss.backward()
```

__________

Optimizer都來自於[torch.optim](https://pytorch.org/docs/stable/optim.html)

首先要去構造它, 把自己設定的模型放進去, 設定learning_rate 並使用step進行更新
```py
import torch
import torchvision
from torch import nn
from torch.utils.data import DataLoader
# from torch.utils.tensorboard import SummaryWriter

dataset = torchvision.datasets.CIFAR10("/CIFAR10", train=False, transform=torchvision.transforms.ToTensor(), download=True)

dataloader = DataLoader(dataset, batch_size=64)

class PJL_NN(nn.Module):
    def __init__(self):
        super(PJL_NN,self).__init__()
        self.model = nn.Sequential(
            nn.Conv2d(3,32,5,padding=2),
            nn.MaxPool2d(2),
            nn.Conv2d(32,32,5,padding=2),
            nn.MaxPool2d(2),
            nn.Conv2d(32,64,5,padding=2),
            nn.MaxPool2d(2),
            nn.Flatten(),
            nn.Linear(1024,64),
            nn.Linear(64,10)
        )

    def forward(self, x):
        x = self.model(x)
        return x

#創建網路
jiun_nn = PJL_NN()

#定義你所用的loss function
loss_fn = nn.CrossEntropyLoss()

#創建優化器
learning_rate = 0.01
optim = torch.optim.SGD(jiun_nn.parameters(), lr=learning_rate)



for data in dataloader:
    imgs, targets = data
    outputs = jiun_nn(imgs)
    loss = loss_fn(outputs, targets)

    #一定要記得把之前算的微分歸零
    optim.zero_grad()

    #此時會更新jiun_nn裡面的nn每層的grad
    loss.backward()

    #對每個參數進行調整
    optim.step()

    print(loss)
```

此時會發現你的loss並無明顯減少, 可能是learning rate或是local minimum,
為了避免上述狀況和overfitting某個dataset, 會對這個data進行多次循環epoch

```py
import torch
import torchvision
from torch import nn
from torch.utils.data import DataLoader
# from torch.utils.tensorboard import SummaryWriter

dataset = torchvision.datasets.CIFAR10("/CIFAR10", train=False, transform=torchvision.transforms.ToTensor(), download=True)

dataloader = DataLoader(dataset, batch_size=64)

class PJL_NN(nn.Module):
    def __init__(self):
        super(PJL_NN,self).__init__()
        self.model = nn.Sequential(
            nn.Conv2d(3,32,5,padding=2),
            nn.MaxPool2d(2),
            nn.Conv2d(32,32,5,padding=2),
            nn.MaxPool2d(2),
            nn.Conv2d(32,64,5,padding=2),
            nn.MaxPool2d(2),
            nn.Flatten(),
            nn.Linear(1024,64),
            nn.Linear(64,10)
        )

    def forward(self, x):
        x = self.model(x)
        return x

#創建網路
jiun_nn = PJL_NN()

#定義你所用的loss function
loss_fn = nn.CrossEntropyLoss()

#創建優化器
learning_rate = 0.01
optim = torch.optim.SGD(jiun_nn.parameters(), lr=learning_rate)


for epoch in range(20):
    running_loss = 0

    for data in dataloader:
        imgs, targets = data
        outputs = jiun_nn(imgs)
        loss = loss_fn(outputs, targets)

        #一定要記得把之前算的微分歸零
        optim.zero_grad()

        #此時會更新jiun_nn裡面的nn每層的grad
        loss.backward()

        #對每個參數進行調整
        optim.step()

        running_loss = running_loss + loss
    
    print(running_loss)

```

**Comment by Pin-Jiun** [comment]
### 調整learning rate- shecduler

```py
import torch
import torchvision
from torch import nn
from torch.utils.data import DataLoader
# from torch.utils.tensorboard import SummaryWriter

dataset = torchvision.datasets.CIFAR10("/CIFAR10", train=False, transform=torchvision.transforms.ToTensor(), download=True)

dataloader = DataLoader(dataset, batch_size=64)

class PJL_NN(nn.Module):
    def __init__(self):
        super(PJL_NN,self).__init__()
        self.model = nn.Sequential(
            nn.Conv2d(3,32,5,padding=2),
            nn.MaxPool2d(2),
            nn.Conv2d(32,32,5,padding=2),
            nn.MaxPool2d(2),
            nn.Conv2d(32,64,5,padding=2),
            nn.MaxPool2d(2),
            nn.Flatten(),
            nn.Linear(1024,64),
            nn.Linear(64,10)
        )

    def forward(self, x):
        x = self.model(x)
        return x

#創建網路
jiun_nn = PJL_NN()

#定義你所用的loss function
loss_fn = nn.CrossEntropyLoss()

#創建優化器
learning_rate = 0.01
optim = torch.optim.SGD(jiun_nn.parameters(), lr=learning_rate)
#使用scheduler更新learning_rate
scheduler = torch.optim.lr_scheduler.StepLR(optim, step_size=30, gamma=0.1)


for epoch in range(20):
    running_loss = 0

    for data in dataloader:
        imgs, targets = data
        outputs = jiun_nn(imgs)
        loss = loss_fn(outputs, targets)

        #一定要記得把之前算的微分歸零
        optim.zero_grad()

        #此時會更新jiun_nn裡面的nn每層的grad
        loss.backward()

        #對每個參數進行調整, 使用scheduler
        optim.step()
        scheduler.step()

        running_loss = running_loss + loss
    
    print(running_loss)
```
